### PR TITLE
Cast model values to the correct PHP type

### PR DESCRIPTION
**Pull request by ausi**
~~Proof of concept.~~

With this we would get the correct types (bool, int or float) depending on the database type for the values in our models.
Together with #4797 we could then cleanup the model property types quite good, see `PageModel` as an example in this PR.

Setting the wrong type is still allowed for now, but triggers a deprecation notice.

Todo:

- [x] ~~Merge #4797 first~~ No: merge this one first!
- [x] Do the same for the `User` class
- [x] Adapt all tests to use the correct types

@contao/developers WDYT?

### CODE REVIEW

**Comment by Toflar** [comment]
Love this! Not sure about the implementation, though. Did you look into the types of Doctrine itself? They all have a `convertToDatabaseValue()` and `convertToPhpValue()` ;)

**Comment by ausi** [comment]
> Did you look into the types of Doctrine itself?

Yes, I tried `convertToPhpValue()` first but it caused more problems than it solved. Because of how our models work, I think applying this only for the simple scalar types is the better way for now.

**Review by fritzmg — COMMENTED** [review]
General concept looks good to me 👍 

Not sure about the cache file location, but I have no better suggestion. Also may be the functionality itself should be its own service, rather than building it into the Model class?

**Comment by ausi** [comment]
> Also may be the functionality itself should be its own service, rather than building it into the Model class?

I thought that performance wise a static array in the `Model` class would be better than having to call `System::getContainer()` all the time 😕

File: core-bundle/src/Cache/ContaoCacheWarmer.php [hunk 1]
```diff
@@ -199,6 +201,14 @@ private function generateTemplateMapper(string $cacheDir): void
         );
     }
 
+    private function generateColumnCastTypes(string $cacheDir): void
+    {
+        $this->filesystem->dumpFile(
+            Path::join($cacheDir, 'contao/config/column_cast_types.php'),
```
Comment (leofeyer) [inline core-bundle/src/Cache/ContaoCacheWarmer.php]: Just a nitpick, I would like `column-types.php` better. 🙈

Comment (ausi) [reply core-bundle/src/Cache/ContaoCacheWarmer.php]: Renamed in 336ec3db7663371f3c6ad74aa6c67d32bd280d5c

**Review by fritzmg — COMMENTED** [review]
Just to confirm: in my PR I will then be able to change all `string|boolean` type hints to `boolean`, correct?

**Comment by ausi** [comment]
> in my PR I will then be able to change all `string|boolean` type hints to `boolean`, correct?

yes

File: core-bundle/src/Resources/contao/models/UserModel.php [hunk 2]
```diff
@@ -45,18 +45,18 @@
  * @property string|array|null $formp
  * @property string|array|null $amg
  * @property string|boolean    $disable
- * @property string|integer    $start
- * @property string|integer    $stop
+ * @property integer           $start
+ * @property integer           $stop
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/models/UserModel.php]: These are `varchar` in the database and thus they would still be returned as a string from there - and not converted to integer automatically, will they?

Comment (ausi) [reply core-bundle/src/Resources/contao/models/UserModel.php]: Fixed in abdd3530a5f4bd2cf3240e7f7634e876ce33b014

Comment (fritzmg) [reply core-bundle/src/Resources/contao/models/UserModel.php]: On a side note: why _do_ we save those timestamps as a string in the first place? I guess it's just an historic remnant from Contao 2 and we could may be change the field type to `int(10) default NULL` and then alter the queries to

```diff
-start='' OR start<='$time'
+start IS NULL OR start<=$time
```

❓

File: calendar-bundle/src/Resources/contao/models/CalendarEventsModel.php [hunk 3]
```diff
@@ -16,50 +16,50 @@
...
+ * @property string|null       $description
+ * @property string            $location
+ * @property string            $address
+ * @property string|null       $teaser
+ * @property string|boolean    $addImage
+ * @property string|boolean    $overwriteMeta
+ * @property string|null       $singleSRC
+ * @property string            $alt
+ * @property string            $imageTitle
+ * @property integer           $size
```
Comment (fritzmg) [inline calendar-bundle/src/Resources/contao/models/CalendarEventsModel.php]: `tl_calendar_events.size` and the other image size fields are a `varchar` and contains either a string consisting of a serialized array or a database ID - so this needs to stay `string|integer` as well.

File: core-bundle/src/Resources/contao/models/FilesModel.php [hunk 4]
```diff
@@ -21,21 +21,21 @@
...
- * @property string              $hash
- * @property string|boolean      $found
- * @property string              $name
- * @property string|float        $importantPartX
- * @property string|float        $importantPartY
- * @property string|float        $importantPartWidth
- * @property string|float        $importantPartHeight
- * @property string|array|null   $meta
+ * @property integer           $id
+ * @property integer|null      $pid
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/models/FilesModel.php]: This seems to have been a previous mistake. `tl_files.pid` is always an UUID or NULL. So this should always have been `string|null` (just like `tl_files.uuid`).

File: core-bundle/src/Resources/contao/models/LayoutModel.php [hunk 5]
```diff
@@ -32,7 +32,7 @@
  * @property string|array|null $modules
  * @property string            $template
  * @property string|boolean    $minifyMarkup
- * @property string|integer    $lightboxSize
+ * @property integer           $lightboxSize
```
Comment (fritzmg) [inline core-bundle/src/Resources/contao/models/LayoutModel.php]: Should be `string|integer` (see above).

**Review by fritzmg — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/models/UserModel.php (hunk 2 above)
Comment (ausi) [reply core-bundle/src/Resources/contao/models/UserModel.php]: Agree, we should change that.
In a separate PR?

Comment (fritzmg) [reply core-bundle/src/Resources/contao/models/UserModel.php]: Yes, definitely a separate PR as this would then be BC breaking change.

**Review by fritzmg — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @ausi.

**Pull request merged**